cmd/createorupdate/update.go

```diff
@@ -11,18 +12,80 @@ import (
 	"github.com/openshift/openshift-azure/pkg/log"
 )
 
-func update(ctx context.Context, cs *api.OpenShiftManagedCluster, p api.Plugin) error {
+func update(ctx context.Context, cs, oldCs *api.OpenShiftManagedCluster, p api.Plugin, azuredeploy []byte) error {
 	config := auth.NewClientCredentialsConfig(ctx.Value(api.ContextKeyClientID).(string), ctx.Value(api.ContextKeyClientSecret).(string), ctx.Value(api.ContextKeyTenantID).(string))
 	authorizer, err := config.Authorizer()
 	if err != nil {
 		return err
 	}
-
 	ssc := compute.NewVirtualMachineScaleSetsClient(cs.Properties.AzProfile.SubscriptionID)
 	ssc.Authorizer = authorizer
 	vmc := compute.NewVirtualMachineScaleSetVMsClient(cs.Properties.AzProfile.SubscriptionID)
 	vmc.Authorizer = authorizer
 
+	// Need to determine whether the current update is a scale operation before
+	// applying the ARM template. For scale up, we need to figure out which are
+	// the new VMs that were created and wait for them to turn ready. For scale
+	// down, we need to drain the correct VMs before applying the ARM template
+	// scales them down.
+	scaleOps := isScaleOp(cs, oldCs)
+	vmsBefore := map[string]struct{}{}
+	for role, scale := range scaleOps {
+		vms, err := listVMs(ctx, cs, vmc, role)
```

Comment:
should we not just always do this for all the roles?

Reply:
the key action here is draining VMs that deploy() is about to cause to be deleted.  A potential case could be after a previously infra/compute upgrade failed where there are currently N+1 VMs all serving workload.
I think it'd be highly unexpected for this code to be doing anything with master VMs - you could just iterate infra and compute roles for now.

Reply:
I think this is the case where storing the diff is invalid - should be storing the number of VMs that is expected post-deploy.

Reply:
By "diff", do you mean the field in `scaleOp` which I just removed in favor of `scaleOp.count` or the diff between new and old VMs that we do now in case of a scale up to figure out which are the new VMs we should wait for? In case of the latter, I am not sure we have any use for the number of desired VMs post-deploy even in the case we are coming from a failed upgrade with a N+1 VMs currently running. `deploy` during the current scale up should either be a no-op (because we scaled up to the actual VMs) or we should scale up past the surge VM. In both cases, the diff between new and old VMs should still work as expected (we should wait for the VMs created during this scale up).